Upgrade immer to 7.0.0

aiwenar opened the pull request:
#### Is this adding or improving a _feature_ or fixing a _bug_?

Upgrading a dependency.

#### What's the new behavior?

Hopefully none.

I'm proposing this upgrade because I've run across immerjs/immer#559, a fix for which is included in immer >= 6.0.4. This issue does not occur in Slate itself (hence why I didn't mark this PR as fixing a bug), but rather when a Slate type such as `Range` is stored in an immer-controlled container, and `transform`ed. For example this:
```typescript
let state = {
    anchor: { path: [0, 0], offset: 0 },
    focus: { path: [0, 0], offset: 10 },
}

state = produce(state, s => Range.transform(s, { type: 'insert_text', path: [0, 0], offset: 5, text: 'new' }))
```
will result in `state.focus.path` being a revoked proxy.

Since this changes a major version of a public dependency it is also a breaking change in Slate. However, I have verified that Slate works with Immer 5, 6, and 7, so in theory we could accept immer >= 5.00 < 8.0.0 and let the consumer choose which version of Immer they want, thus avoiding a breaking change. I'm not proposing that in this PR as I don't know how to do that in npm.

#### Have you checked that...?

- [x] The new code matches the existing patterns and styles.
- [x] The tests pass with `yarn test`.
- [x] The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
- [x] The relevant examples still work. (Run examples with `yarn start`.)

#### Does this fix any issues or need any specific reviewers?

Fixes: 
Reviewers:

aiwenar commented on packages/slate/src/interfaces/node.ts:
```diff
@@ -197,7 +197,7 @@ export const Node = {
     }
 
-    const newRoot = produce(root, r => {
+    const newRoot = produce({ children: root.children }, r => {
```
Since 7.0.0, Immer freezes all properties of the new state (see https://github.com/immerjs/immer/issues/462 for more details). This caused `editor.operations` to be frozen when calling `Editor.fragment(editor, ...)`.

BrentFarese replied:
Confused by this comment.  Does this PR introduce an issue with `editor.operations` by freezing it if `Editor.fragment` is called?

aiwenar replied:
If I recall correctly, in immer 6 the way freezing works has changed. Previously immer (5) would only freeze the new object, starting with 6 it will also recursively freeze all properties. And because of structural sharing, this means that any property of `root` unchanged by the reducer function would also be frozen.

BrentFarese commented:
It looks ok to me.  The 1 thing I worry about is the test coverage in core not covering all areas and some issue with freezing some portion of the editor state creeping up, or some other Immer issue.  But in general, I think we should upgrade this dependency.  It is a major dependency tho.

BrentFarese commented:
Glad to merge this in @aiwenar if you can clean up the conflicts and get it passing CI?  Thanks!

aiwenar commented:
Sure. Would you prefer me to merge master into this branch, or rebase this branch on master?

BrentFarese commented:
> Sure. Would you prefer me to merge master into this branch, or rebase this branch on master?

Either works for me.  Thanks!

aiwenar commented:
Squashed and rebased on top of master, otherwise unchanged. Since this PR was opened immer released versions 7.0.15 and 8.0.0, but I've left it at 7.0.7 as this is the version I've used when I opened this PR.

BrentFarese commented:
Fixes #4034.

niieani commented:
Any ETA on when we can expect a next beta containing this to be released? I'd love to be able to make use of the new TypeScript improvements, but cannot since the last one out (0.61) is broken due to #4034.

Many thanks for your help! 🙇

BrentFarese commented:
> @BrentFarese Any ETA on when we can expect a next beta containing this to be released? I'd love to be able to make use of the new TypeScript improvements, but cannot since the last one out (0.61) is broken due to #4034.
> 
> Many thanks for your help! 🙇

No exact ETA but hope to release something shortly so this issue gets resolved on the `@next` release.

williamstein commented:
I can confirm that @next is updated, and with that update, which includes this PR, #4034 is fixed for me.  Thanks!
